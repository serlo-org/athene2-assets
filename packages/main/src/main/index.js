/**
 * This file is part of Athene2 Assets.
 *
 * Copyright (c) 2017-2019 Serlo Education e.V.
 *
 * Licensed under the Apache License, Version 2.0 (the "License")
 * you may not use this file except in compliance with the License
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * @copyright Copyright (c) 2013-2019 Serlo Education e.V.
 * @license   http://www.apache.org/licenses/LICENSE-2.0 Apache License 2.0
 * @link      https://github.com/serlo-org/athene2-assets for the canonical source repository
 */
/* globals gaOptout, MathJax */
import autosize from 'autosize'
import $ from 'jquery'
import 'jquery-sticky'
import 'jquery-ui'
import 'katex/dist/katex.css'
import 'magnific-popup'
import moment from 'moment'

import { initContent, initEntityEditor } from '../editor'
import '../libs/polyfills'
import Common from '../modules/common'
import Content from '../modules/content'
import '../modules/modals'
import '../modules/spoiler'
import SystemNotification from '../modules/system_notification'
import t from '../modules/translator'
import '../thirdparty/jquery.nestable'
import '../thirdparty/deployggb'
import './libs/easing'
import './libs/event_extensions'
import AjaxOverlay from './modules/ajax_overlay'
import Breadcrumbs from './modules/breadcrumbs'
import { initChangeDimensionEvents } from './modules/change-dimension-events'
import { initContentApi } from './modules/content_api'
import { initConsentBanner } from './modules/consent_banner'
import './modules/forum_select'
import './modules/injections'
import './modules/input_challenge'
import './modules/math_puzzle'
import './modules/math_puzzle/algebra'
import './modules/math_puzzle/touchop'
import MobileNavigation from './modules/mobile_navigation'
import './modules/multiple_choice'
import './modules/profile_birdnest'
import './modules/recaptcha'
import './modules/sentry'
import SideElement from './modules/side_element'
import SideNavigation from './modules/side_navigation'
import './modules/single_choice'
import './modules/slider'
import './modules/sortable_list'
import Supporter from './modules/supporter'
import './modules/timeago'
import './modules/toggle'
import initTracking from './modules/tracking'
import './modules/trigger'
import './styles/main.scss'

window.$ = $
window.jQuery = $

// Needs to be a require call since `window.$` is needed
require('bootstrap-sass')
require('bootstrap-datepicker')
require('jasny-bootstrap/dist/js/jasny-bootstrap.js')

const { version } = require('../../package.json')

console.log('########################')
console.log(`# athene2-assets@${version} #`)
console.log('########################')

const setLanguage = () => {
  const language = $('html').attr('lang') || 'de'

  t.config({
    language
  })

  moment.locale(language)
}

const initNavigation = () => {
  /* eslint-disable no-new */
  new MobileNavigation()
  new Breadcrumbs()
  new SideNavigation()
  /* eslint-enable no-new */
}

const initFooter = () => {
  const $footer = $('#footer')
  const $footerPush = $('#footer-push')
  const $contentLayout = $('#content-layout')
  const $wrap = $('.wrap')
  const $sideContextCourse = $('.side-context-course')

  $footerPush.css('height', $footer.height())
  $wrap.css('margin-bottom', -$footer.height())
  setTimeout(function() {
    $sideContextCourse.css('max-height', $contentLayout.outerHeight())
  }, 300)
  $(window).bind('change-width', function() {
    $footerPush.css('height', $footer.height())
    $wrap.css('margin-bottom', -$footer.height())
    $sideContextCourse.css('max-height', $contentLayout.outerHeight())
    if (MathJax) MathJax.Hub.Queue(['Reprocess', MathJax.Hub])
    $('.nest-statistics').renderNest()
  })
}

const initSubjectNav = $context => {
  let opened = false
  const $dropdown = $('#subject-nav .subject-dropdown')

  $dropdown.click(e => {
    // stop bubbling so that outside click won't close it
    e.stopPropagation()
  })

  const closeDropdown = () => {
    opened = false
    $dropdown.removeClass('open')
    $context.unbind('click', closeDropdown)
  }

  $('.subject-dropdown-toggle', $dropdown).click(e => {
    e.preventDefault()
    opened = !opened
    $dropdown.toggleClass('open', opened)
    if (opened) {
      $context.click(closeDropdown)
    }
  })
}

const init = $context => {
  setLanguage()
  initChangeDimensionEvents()
  initContentApi()
  initContent()
  initConsentBanner()

  // create an system notification whenever Common.genericError is called
  Common.addEventListener('generic error', () => {
    SystemNotification.error()
  })

  if (!$('body').hasClass('serlo-home')) {
    $('#subject-nav').sticky()
  }

  Content.add($context => {
    $('.sortable', $context).SortableList()
    $('.timeago', $context).TimeAgo()
    $('.dialog', $context).SerloModals()
    $('.datepicker', $context).datepicker({
      format: 'yyyy-mm-dd'
    })
    $('.input-daterange', $context).datepicker({
      format: 'yyyy-mm-dd'
    })
    $('.spoiler', $context).Spoiler()
    $('.injection', $context).Injections()
    $('[data-toggle]', $context).ToggleAction()
    $('[data-trigger]', $context).TriggerAction()
    $('form[name="discussion"]', $context).ForumSelect()
    $('.carousel.slide.carousel-tabbed', $context).Slider()
    $('.nest-statistics', $context).renderNest()
    $('.math-puzzle', $context).MathPuzzle()
    $('form:has(button.g-recaptcha)').ReCaptcha()
    $('.ory-edit-button', $context).click(function(e) {
      e.preventDefault()
      initEntityEditor(
        $(this).data('id'),
        $(this).attr('href'),
        $(this).data('type')
      )
      $('.convert-button').hide()
    })

    $('.convert-button', $context).click(function(e) {
      e.preventDefault()
      const id = $(this).data('content-id')
      const href = $(this).attr('href')
      const $loading = $('#loading')

      $.ajax({
        url: href,
        type: 'GET',
        async: true,
        beforeSend: () => {
          $loading.show()
        }
      }).done(function(data) {
        function getEditedArticle($all) {
          const $editable = $all.find(
            `.editable[data-id="${id}"][data-edit-type="ory"]`
          )
          return $editable.closest('article').length
            ? $editable.closest('article')
            : $all.find('#content-layout article')
        }

        const $target = getEditedArticle($('body'))
        const $dataArticle = getEditedArticle($(data))
        $target.html($dataArticle.html())

        const $editButton = $(`.ory-edit-button[data-id="${id}"]`, data)
        initEntityEditor(
          id,
          $editButton.attr('href'),
          $editButton.data('type')
        ).then(() => {
          Common.trigger('new context', $target)
          $loading.hide()
          $('.convert-button').hide()
        })
      })
    })

    if ($('#ory-edit-form', $context).length > 0) {
      initEntityEditor(
        $('#ory-edit-form', $context).data('id'),
        $('#ory-edit-form form', $context).attr('action') ||
          window.location.pathname,
        $('#ory-edit-form', $context).data('type')
      )
      $('.convert-button').hide()
    }
    // Dirty Hack for Course Pages Mobile
    if ($('.side-context-course').length > 0) {
      $('#content-layout').addClass('course-page')
    }
    $('.text-exercise:has(.input-challenge-group)', $context).InputChallenge()
    $('.text-exercise:has(.single-choice-group)', $context).SingleChoice()
    $('.text-exercise:has(.multiple-choice-group)', $context).MultipleChoice()
    autosize($('textarea.autosize'))
    $('.r img', $context).each(function() {
      var $that = $(this)
      $that.magnificPopup({
        type: 'image',
        closeOnContentClick: true,
        fixedContentPos: false,
        items: {
          src: $that.attr('src') || $that.data('cfsrc') // CloudFlare Mirage lazy loading
        },
        image: {
          verticalFit: true
        },
        disableOn: function() {
          return $that.parents('a').length <= 0
        }
      })
    })
    if (MathJax) MathJax.Hub.Queue(['Typeset', MathJax.Hub])
  })

  // Tooltips opt in
  $('[data-toggle="tooltip"]').tooltip({
    container: 'body'
  })
  Common.addEventListener('new context', Content.init)

  initNavigation()
  initFooter()

  // initialize ajax overlay
  let ajaxOverlay
  ajaxOverlay = new AjaxOverlay({
    on: {
      contentOpened: function() {
        Content.init(this.$el)
      },
      error: function(err) {
        ajaxOverlay.shutDownAjaxContent()
        SystemNotification.error(
          t(
            'When asynchronously trying to load a ressource, I came across an error: %s',
            err.status + ' ' + err.statusText
          )
        )
      }
    }
  })

  Common.trigger('new context', $context)

  initSubjectNav($context)

  SideElement.init()

  initTracking($context)
}

init($('body'))
Supporter.check()
