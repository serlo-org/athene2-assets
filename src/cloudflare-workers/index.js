addEventListener('fetch', event => {
  event.respondWith(handleRequest(event.request))
})

export async function handleRequest (request) {
  return (
    (await handleSemanticAssetsFilenames(request)) || (await fetch(request))
  )
}

async function handleSemanticAssetsFilenames (request) {
  const re = /^https:\/\/assets.serlo.org\/(legacy\/|)((?!legacy)\w+)\/([\w\-\+]+)\.(\w+)$/
  const match = request.url.match(re)

  if (!match) {
    return null
  }

  const [_url, prefix, hash, name, extension] = match
  return await fetch(
    `https://assets.serlo.org/${prefix}${hash}.${extension}`,
    request
  )
}
