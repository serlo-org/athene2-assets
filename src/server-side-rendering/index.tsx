import express from 'express'
import cors from 'cors'
import bodyParser from 'body-parser'
import * as ReactDOMServer from 'react-dom/server'
import * as React from 'react'

interface Registry {
  [name: string]: React.ComponentType<any>
}

export const createApp = (registry: Registry) => {
  const app = express()

  app.use(cors())
  app.use(bodyParser.json())

  app.post('/', (req, res) => {
    const { name, props } = req.body
    const Component = registry[name]
    const html = ReactDOMServer.renderToString(<Component {...props} />)
    res.send({ html })
  })
  return app
}
