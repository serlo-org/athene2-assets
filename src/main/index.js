/* globals gaOptout, MathJax */
import autosize from 'autosize'
import $ from 'jquery'
import 'jquery-sticky'
import 'jquery-ui'
import 'magnific-popup'
import moment from 'moment'
import _ from 'underscore'

import 'cookieconsent'

import '../libs/polyfills'
import Common from '../modules/common'
import Content from '../modules/content'
import '../modules/injections'
import '../modules/modals'
import '../modules/spoiler'
import Supporter from '../modules/supporter'
import SystemNotification from '../modules/system_notification'
import '../modules/timeago'
import t from '../modules/translator'
import '../thirdparty/jquery.nestable'
import '../thirdparty/deployggb'
import './libs/easing'
import './libs/event_extensions'
import AjaxOverlay from './modules/serlo_ajax_overlay'
import Breadcrumbs from './modules/serlo_breadcrumbs'
import './modules/serlo_forum_select'
import './modules/serlo_input_challenge'
import './modules/serlo_math_puzzle/serlo_math_puzzle'
import './modules/serlo_math_puzzle/serlo_math_puzzle_algebra'
import './modules/serlo_math_puzzle/serlo_math_puzzle_touchop'
import MobileNavigation from './modules/serlo_mobile_navigation'
import './modules/serlo_multiple_choice'
import './modules/serlo_profile_birdnest'
import SideElement from './modules/serlo_side_element'
import SideNavigation from './modules/serlo_side_navigation'
import './modules/serlo_single_choice'
import './modules/serlo_slider'
import './modules/serlo_sortable_list'
import './modules/serlo_toggle'
import initTracking from './modules/serlo_tracking'
import './modules/serlo_trigger'
// FIXME historyjs; not needed?

import './styles/main.scss'

window.$ = $
window.jQuery = $

// Needs to be a require call since `window.$` is needed
require('bootstrap-sass')
require('bootstrap-datepicker')
// FIXME needed?
require('jasny-bootstrap/dist/js/jasny-bootstrap.js')

const setLanguage = () => {
  const language = $('html').attr('lang') || 'de'

  t.config({
    language
  })

  moment.locale(language)
}

const initResizeEvent = () => {
  const $window = $(window)
  let cachedWidth = $window.width()

  // `resizeDelay` will be triggered if it wasn't triggered for 0.5s
  $window.resize(
    _.debounce(function () {
      const width = $window.width()
      if (cachedWidth !== width) {
        $(this).trigger('resizeDelay')
        cachedWidth = width
      }
    }, 500)
  )
}

const initNavigation = () => {
  /* eslint-disable no-new */
  new MobileNavigation()
  new Breadcrumbs()
  new SideNavigation()
  /* eslint-enable no-new */
}

const initFooter = () => {
  const $footer = $('#footer')
  const $footerPush = $('#footer-push')
  const $contentLayout = $('#content-layout')
  const $wrap = $('.wrap')
  const $sideContextCourse = $('.side-context-course')

  $footerPush.css('height', $footer.height())
  $wrap.css('margin-bottom', -$footer.height())
  setTimeout(function () {
    $sideContextCourse.css('max-height', $contentLayout.outerHeight())
  }, 300)
  $(window).bind('resizeDelay', function () {
    $footerPush.css('height', $footer.height())
    $wrap.css('margin-bottom', -$footer.height())
    $sideContextCourse.css('max-height', $contentLayout.outerHeight())
    MathJax.Hub.Queue(['Reprocess', MathJax.Hub])
    $('.nest-statistics').renderNest()
  })
}

const initSubjectNav = $context => {
  let opened = false
  const $dropdown = $('#subject-nav .subject-dropdown')

  $dropdown.click(e => {
    // stop bubbling so that outside click won't close it
    e.stopPropagation()
  })

  const closeDropdown = () => {
    opened = false
    $dropdown.removeClass('open')
    $context.unbind('click', closeDropdown)
  }

  $('.subject-dropdown-toggle', $dropdown).click(e => {
    e.preventDefault()
    opened = !opened
    $dropdown.toggleClass('open', opened)
    if (opened) {
      $context.click(closeDropdown)
    }
  })
}

const initContentApi = () => {
  const $page = $('#header, #page')
  const $links = $page.find('a')

  const { href, origin, pathname } = window.location
  const query = href.substr(origin.length + pathname.length)

  $links.each(function () {
    const $link = $(this)
    const url = $link.attr('href')
    const target = $link.attr('target')

    const isInternalLink =
      url && (url.startsWith('/') || url.startsWith(origin))
    const isBlank = target && target === '_blank'

    if (isInternalLink && !isBlank) {
      $link.attr('href', `${url}${query}`)
    }
  })
}

const initConsentBanner = () => {
  window.cookieconsent.initialise({
    content: {
      message:
        'Mit der Nutzung dieser Webseite erklärst Du Dich damit einverstanden, dass wir Cookies verwenden.',
      dismiss: 'Verstanden',
      link: 'Mehr Informationen findest Du in unserer Datenschutzerklärung',
      href: 'https://de.serlo.org/datenschutz'
    },
    type: 'info',
    palette: {
      popup: { background: '#007ec1', text: '#ffffff', link: '#ffffff' },
      button: { background: 'transparent', border: '#ffffff', text: '#ffffff' }
    }
  })
}

const init = $context => {
  setLanguage()
  initResizeEvent()
  initContentApi()
  initConsentBanner()

  // create an system notification whenever Common.genericError is called
  Common.addEventListener('generic error', () => {
    SystemNotification.error()
  })

  if (!$('body').hasClass('serlo-home')) {
    $('#subject-nav').sticky()
  }

  Content.add($context => {
    $('.sortable', $context).SortableList()
    $('.timeago', $context).TimeAgo()
    $('.dialog', $context).SerloModals()
    $('.datepicker', $context).datepicker({
      format: 'yyyy-mm-dd'
    })
    $('.input-daterange', $context).datepicker({
      format: 'yyyy-mm-dd'
    })
    $('.spoiler', $context).Spoiler()
    $('.injection', $context).Injections()
    $('[data-toggle]', $context).ToggleAction()
    $('[data-trigger]', $context).TriggerAction()
    $('form[name="discussion"]', $context).ForumSelect()
    $('.carousel.slide.carousel-tabbed', $context).Slider()
    $('.nest-statistics', $context).renderNest()
    $('.math-puzzle', $context).MathPuzzle()
    // Dirty Hack for Course Pages Mobile
    if ($('.side-context-course').length > 0) {
      $('#content-layout').addClass('course-page')
    }
    $('.text-exercise:has(.input-challenge-group)', $context).InputChallenge()
    $('.text-exercise:has(.single-choice-group)', $context).SingleChoice()
    $('.text-exercise:has(.multiple-choice-group)', $context).MultipleChoice()
    autosize($('textarea.autosize'))
    $('.r img', $context).each(function () {
      var $that = $(this)
      $that.magnificPopup({
        type: 'image',
        closeOnContentClick: true,
        fixedContentPos: false,
        items: {
          src: $that.attr('src')
        },
        image: {
          verticalFit: true
        },
        disableOn: function () {
          return $that.parents('a').length <= 0
        }
      })
    })
    MathJax.Hub.Queue(['Typeset', MathJax.Hub])
  })

  // Tooltips opt in
  $('[data-toggle="tooltip"]').tooltip({
    container: 'body'
  })
  Common.addEventListener('new context', Content.init)

  initNavigation()
  initFooter()

  // initialize ajax overlay
  let ajaxOverlay
  ajaxOverlay = new AjaxOverlay({
    on: {
      contentOpened: function () {
        Content.init(this.$el)
      },
      error: function (err) {
        ajaxOverlay.shutDownAjaxContent()
        SystemNotification.error(
          t(
            'When asynchronously trying to load a ressource, I came across an error: %s',
            err.status + ' ' + err.statusText
          )
        )
      }
    }
  })

  Common.trigger('new context', $context)

  initSubjectNav($context)

  SideElement.init()

  // Google Analytics opt out
  $('a[href=ga-opt-out]').click(function (e) {
    e.preventDefault()
    gaOptout()
    SystemNotification.notify(
      t('Successfully deactivated Google Analytics'),
      'success'
    )
    window.scrollTo(0, 0)
  })
  initTracking($context)
}

init($('body'))
Supporter.check()
