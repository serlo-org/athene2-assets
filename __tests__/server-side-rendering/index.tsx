/**
 * @jest-environment node
 */

import axios from 'axios'
import * as React from 'react'
import { createApp } from '../../src/server-side-rendering'

const Re = props => {
  return <div>{props.name}</div>
}
const app = createApp({ Re: Re })
describe('asdf', () => {
  let server
  beforeAll(() => {
    return new Promise(resolve => {
      server = app.listen(3000, resolve)
    })
  })
  afterAll(() => {
    server.close()
  })
  test('name', async () => {
    const response = await axios.post('http://localhost:3000', {
      name: 'Re',
      props: { name: 'asdf' }
    })
    const { data } = response
    expect(data.html).toMatchSnapshot()
  })
})
