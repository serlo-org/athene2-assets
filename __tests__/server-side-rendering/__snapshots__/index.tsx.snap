// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asdf name 1`] = `"<div data-reactroot=\\"\\">asdf</div>"`;
